### ISSUE TITLE
Force models to return JSON instead of free-form text

### ISSUE THREAD

**Issue by abhinavkulkarni** [paul-gauthier/aider#263]
Hi,

I have been experimenting with `aider` for the past few days and have been impressed. I have a local [HuggingFace TGI](https://github.com/huggingface/text-generation-inference) installation running Code Llama and a [LiteLLM wrapper](https://docs.litellm.ai/docs/providers/huggingface#text-generation-interface-tgi---llms) that makes it compatible with OpenAI API. 

However, it is a little harder to make it follow the various prompts for editing, creating new files, etc. defined in the `aider` repo. The Code Llama model response often has a preamble and appendix of explanatory text such as, _"Sure, here's a new Python file that prints hello world..."_.

Is it possible to instead return JSON with specified keys and values so that it is easier to parse the LLM output? A lot of API services/frameworks have the ability to provide custom grammar/regex which are nothing but custom `LogitProcessor`s that bias logits so that the model response conforms to the grammar. For example, recently Llama.cpp added [grammar-based sampling](https://twitter.com/ggerganov/status/1698359882234142776?s=20).

OpenAI models are already very good at following instructions and returning JSON outputs due to their superior RLHF training. In the light of this, can we have force the models to return JSON responses instead of free-form text?

Thanks for the great work!

**Comment by tmm1** [comment]
See https://aider.chat/docs/benchmarks.html

**Comment by paul-gauthier** [comment]
I'm going to close this issue for now, but feel free to re-open or file a new issue any time.

**Comment by lee-b** [comment]
Not sure why this one was closed, the answer doesn't seem to fit the issue, and grammar constraint would probably make aider around 3x to 5x more reliable.

**Comment by paul-gauthier** [comment]
Did you read the article? It discusses extensive benchmarking that showed wrapping code in json (function calls) was harmful to the overall ability of the LLM to successfully solve coding problems.

That work is about a year old, and newer models have grammar constraints. I recently did some new benchmarking which seems to show the a similar result. It's on my list to do a more thorough benchmark of this again soon.

**Comment by lee-b** [comment]
I would expect much better results with a grammar that constrains the llm to selecting a tool, and populating appropriate fields for that tool with appropriate values (rather than ANY freeform output except in text fields), with quite surgical tools + tests / lints.

I'm toying with the idea of writing something similar to aider with strict constrained grammars and pydantic-based typing/tools plugins to test this, but would be happy to not have to :)

If you have any preliminary results from something like that, I'd be interested.  Certainly very interested if you produce benchmarks similar to the article but with constraints.

Do agree that newer models should be (somewhat) better at this, either way.  It's probably related to model size too.
